### PR TITLE
Changed from static list of DataSourceFactory options to a DataSourceFactoryRegistry

### PR DESCRIPTION
**Pull request by chad-moller-target**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,39 +57,54 @@ dependencies {
...
-    password = "password",
-    databaseName = "dbName"
-)).dataSource()
+val config = DbConfig(
+  type = "H2_INMEM",
+  username = "user",
+  password = "password",
+  databaseName = "dbName"
+)
+val datasource = DataSourceFactoryRegistry.dataSource(config)
```
Comment (dtanner) [inline README.md:74]: Seems like from an external point of view it's still a `DataSourceFactory`. i.e. It's a factory creating a datasource for you.  Maybe the registry is an internal/private concept.?

File: README.md [hunk 2]
```diff
@@ -57,39 +57,54 @@ dependencies {
...
+See `DbConfig` for a full list of configuration options available.
+
+## Custom Database Types
 
-If another implementation of DataSource is required, you can construct your own as shown below:
+If another implementation of DataSource is required, you can register a custom "Type" to be set on `DbConfig` and build the 
+respective `DataSource` in a `DataSourceFactory` lambda as shown below.
+
+```kotlin
+DataSourceFactoryRegistry.registerDataSourceFactory("custom") {config: DbConfig ->
```
Comment (dtanner) [inline README.md:98]: `Type` is what naturally came to my mind, and then I hunted around your comments and saw you said "you can register a custom Type to be set..."
```suggestion
DataSourceFactory.addType("custom") {config: DbConfig ->
```

File: README.md [hunk 3]
```diff
@@ -57,39 +57,54 @@ dependencies {
...
+val config = DbConfig(
+  type = "H2_INMEM",
+  username = "user",
+  password = "password",
+  databaseName = "dbName"
+)
+val datasource = DataSourceFactoryRegistry.dataSource(config)
 
-val db = Db(dataSource)
+val db = Db(datasource)
```
Comment (dtanner) [inline README.md:76]: `DataSource` or `Datasource` everywhere? Looks like we have a mix.

File: src/main/kotlin/com/target/liteforjdbc/HikariDatasourceFactory.kt [hunk 4]
```diff
@@ -9,21 +9,24 @@ const val DATABASE_HOST_ENV_NAME = "DATABASE_HOST"
...
-    fun dataSource(jdbcUrl: String): HikariDataSource {
-        val dataSource = HikariDataSource()
-        dataSource.jdbcUrl = jdbcUrl
-        dataSource.connectionTimeout = config.connectionTimeoutMillis
-        dataSource.keepaliveTime = config.keepAliveTime
-        dataSource.maxLifetime = config.maxLifetime
-        dataSource.idleTimeout = config.idleTimeoutMillis
-        dataSource.minimumIdle = config.minimumIdle
-        dataSource.maximumPoolSize = config.maximumPoolSize
+fun hikariDbConfigValidate(config: DbConfig) {
```
Comment (dtanner) [inline src/main/kotlin/com/target/liteforjdbc/HikariDatasourceFactory.kt:13]:
```suggestion
fun validate(config: DbConfig) {
```

File: src/main/kotlin/com/target/liteforjdbc/PosgresDatasourceFactory.kt [hunk 5]
```diff
@@ -1,27 +1,27 @@
 package com.target.liteforjdbc
```
Comment (dtanner) [inline src/main/kotlin/com/target/liteforjdbc/PosgresDatasourceFactory.kt:1]: just noticed `Posgres` in this class name is spelled wrong

**Review by dtanner — COMMENTED** [review]
this is really nice stuff. a couple change suggestions around naming mostly

File: README.md [hunk 6]
```diff
@@ -57,39 +62,54 @@ dependencies {
 
 The core of lite-for-jdbc is the Db class. A Db object is intended to be used as a singleton and
 injected as a dependency in repository classes. It requires a DataSource constructor argument,
-and there is a DataSourceFactory to help with that.
+and there is a DataSourceFactoryRegistry to help with that.
```
Comment (dtanner) [inline README.md:65]:
```suggestion
and there is a DataSourceFactory to help with that.
```

File: README.md [hunk 7]
```diff
@@ -174,7 +197,8 @@ val adminUsers: List<User> = db.findAll(
 ```
 
 ## executeUpdate
-```kotlin
```
Comment (dtanner) [inline README.md:177]: why the removal of the kotlin language hint?

File: src/test/kotlin/com/target/liteforjdbc/DataSourceFactoryRegistryTest.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
+package com.target.liteforjdbc
```
Comment (dtanner) [inline src/test/kotlin/com/target/liteforjdbc/DataSourceFactoryRegistryTest.kt:1]: file name - remove Registry

**Review by dtanner — DISMISSED** [review]
a few minor things - looks good

File: README.md (hunk 7 above)
Comment (chad-moller-target) [reply README.md:177]: Because it's just the signature and not a full method it comes up as compilation errors in IntelliJ.

Comment (dtanner) [reply README.md:177]: ah k 👍

**Review by dtanner — APPROVED** [review]

**Pull request merged**